cassiopeia/data/CassiopeiaTree.py

```diff
@@ -68,117 +73,291 @@ class CassiopeiaTree:
+                ] = self.__original_character_matrix.loc[n].to_list()
+            else:
+                self.__network.nodes[n]["character_states"] = []
+
+        # instantiate branch lengths
+        for u, v in self.edges:
+            self.__network[u][v]["length"] = 1
+
+        # instantiate node time
+        self.__network.nodes[self.root]["time"] = 0
+        for u, v in self.depth_first_traverse_edges(source=self.root):
+            self.__network.nodes[v]["time"] = (
+                self.__network.nodes[u]["time"] + self.__network[u][v]["length"]
+            )
+
+    def initialize_character_states_at_leaves(
+        self, character_matrix: Union[pd.DataFrame, Dict]
+    ):
+        """Populates character states at leaves.
 
-        if self.__network is not None:
-
-            # add character states
-            for n in self.nodes:
-                if n in self.character_matrix.index.values:
-                    self.__network.nodes[n]['character_states'] = self.character_matrix.loc[n].to_list()
-                else:
-                    self.__network.nodes[n]['character_states'] = []
-            
-            # instantiate branch lengths
-            for u,v in self.edges:
-                self.__network[u][v]['length'] = 1
-
-            # instantiate node ages and edge depth
-            self.__network.nodes[self.root]['age'] = 0
-            self.__network.nodes[self.root]['edge_depth'] = 0
-            for u,v in self.depth_first_traverse_edges(source=self.root):
-                self.__network.nodes[v]['age'] = self.__network.nodes[u]['age'] + self.__network[u][v]['length']
-                self.__network.nodes[v]['edge_depth'] = self.__network.nodes[v]['age']
+        Assigns character states to the leaves of the tree. This function
+        must have a character state assignment to all leaves of the tree.
+
+        Args:
+            character_matrix: A pandas dataframe or dictionary for mapping
+                character states to the leaves of the tree.
+        
+        Raises:
+            CassiopeiaTreeError if not all leaves are accounted for or if the
+                tree has not been initialized.
+        """
+        if self.__network is None:
+            raise CassiopeiaTreeError("Tree has not been initialized.")
+
+        if isinstance(character_matrix, dict):
+            character_matrix = pd.DataFrame.from_dict(
+                character_matrix, orient="index"
+            )
+
+        if len(np.setdiff1d(self.leaves, character_matrix.index.values)) > 0:
+            raise CassiopeiaTreeError(
+                "Character matrix does not account for all the leaves."
+            )
+
+        for n in self.leaves:
+            self.__set_character_states(n, character_matrix.loc[n].tolist())
+
+        self.__original_character_matrix = character_matrix.copy()
+        self.__current_character_matrix = character_matrix.copy()
+
+    def initialize_all_character_states(self, mapping: Dict):
```

Comment:
I thnk the naming should be more explicit here too, like "initialize_character_matrix_from_mapping". Also, the docstring here does not seem to match the code as the code only labels leaves.

Reply:
For this specific method, the states of internal nodes are also initialized, so this method is more encompassing than just the character matrix (which only holds data for leaves).

Reply:
Ah I see. whoops read this too fast!